@import ix.core.search.TextIndexer.Facet
@import ix.idg.controllers.IDGApp
@import ix.idg.controllers.Commons
@import java.net.URLEncoder

@(title: String, target: Facet, disease: Facet, ligand: Facet)

@fcol(name: String, facet: Facet) = {
    <td>
    @if(facet != null) {
        @for(f <- facet.getValues()) {
            @if(name.equals(f.getLabel())) {
                @f.getCount()
            }
        }
    }
    </td>
}

<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
    <head>
        <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
        <meta http-equiv="X-UA-Compatible" content="IE=edge">

        <title>@title</title>
        <link rel="stylesheet"
        href='@routes.WebJarAssets.at(WebJarAssets.locate("bootstrap.min.css"))'>
        <link rel="stylesheet" type="text/css"
        href='@routes.WebJarAssets.at(WebJarAssets.locate("font-awesome.min.css"))'>
        <link rel="stylesheet" type="text/css"
        href='@routes.Assets.versioned("ncats/css/idg-landing-hex.css")'>
        <link rel="icon" type="image/png" href='@routes.Assets.versioned("images/favicon.png")'>
        <style type="text/css">

            .lead {
                font-weight:bolder;
            }

        h5 {
        color : #666666 ;
        display : inline ;
        line-height : 1.75 ;
        font-style : italic ;
        padding-left: 1em;
        }

        </style>
            <!--[if lte IE 9]>
<style type="text/css">
#ie6msg {border:3px solid #c33; margin:8px 0; background:#fcc; color:#000;}
#ie6msg h4{margin:8px; padding:0;}
#ie6msg p{margin:8px; padding:0;}
#ie6msg p a.getie7{font-weight:bold; color:#006;}
#ie6msg p a.ie6expl{font-weight:bold; color:#006;}
</style>
<div id="ie6msg">
<h4>Did you know that your browser is out of date, and potentially poses a security risk?</h4>
<p>To get the best possible experience using this website it is recommended that you upgrade your browser to a newer version.
<p>You could also try some other popular browsers like <a class="ie6expl" href="http://mozilla.com">FireFox</a> or
 <a class="ie6expl" href="https://www.google.com/chrome/S">Chrome</a>.</p>
</div>
<![endif]-->
        <script src='@routes.WebJarAssets.at(WebJarAssets.locate("jquery.min.js"))'
        type='text/javascript'></script>
        <script src='@routes.WebJarAssets.at(WebJarAssets.locate("bootstrap.min.js"))'></script>
        <script src='@routes.WebJarAssets.at(WebJarAssets.locate("typeahead.bundle.min.js"))'></script>
        <script src='@routes.WebJarAssets.at(WebJarAssets.locate("handlebars.min.js"))'></script>
        <script src='@routes.WebJarAssets.at(WebJarAssets.locate("bloodhound.min.js"))'></script>
        <script>
        ( function ( i, s, o, g, r, a, m ) { i[ 'GoogleAnalyticsObject' ] = r ; i[ r ] = i[ r ] || function ( ) {
        ( i[ r ].q = i[ r ].q ||[ ] ).push ( arguments ) }, i[ r ].l = 1 * new Date ( ) ; a = s.createElement ( o ),
        m = s.getElementsByTagName ( o )[ 0 ] ; a.async = 1 ; a.src = g ; m.parentNode.insertBefore ( a, m )
        } ) ( window, document, 'script', '//www.google-analytics.com/analytics.js', 'ga' ) ;

        ga ( 'create', 'UA-53774738-1', 'auto' ) ;
        ga ( 'send', 'pageview' ) ;
        </script>
    </head>
    <body>
        <div class="jumbotron">
            <div class="container">
                <center><h1>About Pharos</h1></center>
            </div>
        </div>

        <div class="container vertical-center" id="main-box">
            <div class="row">
                <p>Pharos is the user interface to the <a href="http://targetcentral.ws/">Knowledge Management Center</a>
                    (KMC) for the <a href="https://commonfund.nih.gov/idg/index">Illuminating the Druggable Genome</a>
                    (IDG) program funded by the National Institutes of Health (NIH) Common Fund.
                    (Grant No. <a href="https://projectreporter.nih.gov/project_info_details.cfm?aid=8898033&icde=28628127&ddparam=&ddvalue=&ddsub=&cr=3&csb=default&cs=ASC">
                        5U54CA189205-02</a>). The goal of KMC is to develop a comprehensive, integrated knowledge-base
                    for the Druggable Genome (DG) to illuminate the uncharacterized and/or poorly annotated portion of
                    the DG, focusing on four of the most commonly drug-targeted protein families: &nbsp;G-protein-coupled
                    receptors (GPCRs); nuclear receptors (NRs); ion channels (ICs); and kinases.
                </p>
                <p>Based on modern web design principles the interface provides facile access to all data types
                    collected by the KMC.
                    Given the complexity of the data surrounding any target, efficient and intuitive visualization has
                    been a high priority, to enable users to quickly navigate  &amp; summarize search results and
                    rapidly identify patterns. A critical feature of the interface is the ability to perform flexible
                    search and subsequent drill down of search results. Underlying the interface is a RESTful API that
                    provides programmatic access to all KMC data, allowing for easy consumption in user applications.
                </p>
                <p>The <a href="">Help</a> page provides more information on various topics.</p>
                <p>For feedback and comments, please contact us at
                    <a href="mailto:pharos@@mail.nih.gov"><code>pharos@@mail.nih.gov</code></a>
                </p>

            </div>
            <div class="row">
                <p class="lead">Who's involved?</p>
                <p>Pharos is developed at <a href="http://ncats.nih.gov">NCATS</a>, together with collaborators from
                    the <a href="http://hsc.unm.edu/">University of New Mexico</a>, <a href="http://icahn.mssm.edu/">
                    Icahn School of Medicine, Mount Sinai</a>, <a href="https://www.ebi.ac.uk/about">EMBL-EBI</a>,
                    the <a href="http://www.cpr.ku.dk/">Novo Nordisk Foundation Center for Protein Research (U. Copenhagen)</a>
                    and the <a href="http://ccs.miami.edu/?p=1103">University of Miami</a>.
                    </p>
                <p>
                    The IDG TechDev network is the experimental side of IDG. TechDev research aims to develop and
                    utilize technologies and streamlined experimental workflows for functional studies of poorly
                    characterized and/or un-annotated genes and proteins, thereby providing high value data and knowledge
                    for KMC.  KMC in turn is tasked with providing guidance as to research priorities based on knowledge
                    gaps and druggability likelihood analyses.
                </p>
                <p>
                    To find more details on our collaborators visit <a href="http://targetcentral.ws/">http://targetcentral.ws/</a>
                </p>
            </div>
            <div class="row">
                <p class="lead">What data is available?</p>
                <p>The data available in Pharos is obtained from the <a href="http://juniper.health.unm.edu/tcrd/">
                    Target Central Resource Database</a> which integrates data from a variey of data sources including
                    the <a href="http://amp.pharm.mssm.edu/Harmonizome/">Harmonizome</a>, Jensen Lab datasets, EBI data
                    sets (such as <a href="https://www.ebi.ac.uk/chembl/">ChEMBL</a>) and the
                    <a href="http://drugtargetontology.org/about.html">Drug Target Ontology</a> (DTO)
                    from U. Miami. TCRD integration methodology involves
                    importation by value or by reference to external sources, as informed by performance, provenance,
                    and other design criteria. See <a href="http://targetcentral.ws/Pharos">here</a> for a full listing
                    of datasets incorporated in to <a href="http://juniper.health.unm.edu/tcrd/">TCRD</a>.
                </p>
                <p>
                    The key data types represented in Pharos are listed below:
                </p>

                <h5>Small molecule data</h5> including approved drug data, bioassay data
                <br>
                <h5>Protein data</h5> including protein-protein interaction data
                <br>
                <h5>Disease data</h5> from <a href="http://www.ncbi.nlm.nih.gov/omim">OMIM</a> and
                <a href="http://disease-ontology.org/">Disease Ontology</a>
                <br>
                <h5>Genomic data</h5> including expression (protein, RNA), transcription factors and epigenomic
                associations
                <br>
                <h5>Phenotypic data</h5> including mouse phenotypes, mouse/human orthologs and GWAS results
                <br>
                <h5>Funding data</h5> via NIH <a href="https://projectreporter.nih.gov/reporter.cfm">RePORTER</a>
                <br>
                <h5>Text data</h5> including <a href="http://www.ncbi.nlm.nih.gov/gene/about-generif">GeneRIF</a>'s
                and text-mined publications
                <br>
                <h5>Ontologies</h5> including the Drug Target Ontology, <a href="http://disease-ontology.org/">Disease
                Ontology</a>, <a href="http://www.pantherdb.org/panther/ontologies.jsp">PANTHER</a> and
                <a href="http://geneontology.org/">GO</a>.
                <p></p>
            </div>
            <div class="row">
                <p class="lead">What license does Pharos use?</p>
                <p>All data accessed from Pharos is made available under the <a href="https://creativecommons.org/licenses/by-sa/4.0/">
                    CC BY-SA 4.0</a> license.</p>
            </div>
            <div class="row">
                <p class="lead">Where can I get the code?</p>
                <p>The sources for the Pharos web interface are available from <a href="https://spotlite.nih.gov/ncats/pharos">
                    https://spotlite.nih.gov/ncats/pharos</a>. The repository provides
                    <a href="https://spotlite.nih.gov/ncats/pharos/tree/master">instructions</a> on building and installation</p>
            </div>
            <div class="row">
                <p class="lead">How can I cite Pharos?</p>
                <p>If you use Pharos, please consider citing it as
                    <blockquote>Nguyen, D.-T. et al, "Pharos: Shedding Light on the Druggable Genome", <i>submitted</i></blockquote>
                </p>
            </div>

            @*<div class="row">*@
            @*<div class="col-md-6 col-md-offset-3">*@
            @*<div class="panel panel-default">*@
            @*<div class="panel-body">*@
            @*<table class="table table-striped">*@
            @*<tr>*@
            @*<th>Source</th>*@
            @*<th>Target</th>*@
            @*<th>Disease</th>*@
            @*<th>Ligand</th>*@
            @*</tr>*@
            @* @for(source <- IDGApp.getDataSources()) {*@
            @*<tr>*@
            @* @if(source.href != null) {*@
            @*<td><a href="@source.href">@source.name</a></td>*@
            @*} else {*@
            @*<td>@source.name</td>*@
            @*}*@
            @*<td><a href='@{*@
            @*ix.idg.controllers.routes.IDGApp.targets() + "?facet=" + URLEncoder.encode(Commons.SOURCE, "utf8") + "/" + URLEncoder.encode(source.name, "utf8")*@
            @*}'>@source.targets</a>*@
            @*</td>*@
            @*<td><a href='@{*@
            @*ix.idg.controllers.routes.IDGApp.diseases() + "?facet=" + URLEncoder.encode(Commons.SOURCE, "utf8") + "/" + URLEncoder.encode(source.name, "utf8")*@
            @*}'>@source.diseases</a>*@
            @*</td>*@
            @*<td><a href='@{*@
            @*ix.idg.controllers.routes.IDGApp.ligands() + "?facet=" + URLEncoder.encode(Commons.SOURCE, "utf8") + "/" + URLEncoder.encode(source.name, "utf8")*@
            @*}'>@source.ligands</a>*@
            @*</td>*@
            @*</tr>*@
            @*}*@
            @*</table>*@
            @*</div>*@
            @*</div>*@
            @*</div>*@
            @*</div>*@
        </div>
    </body>
</html>
