@import ix.core.search.TextIndexer.Facet
@import ix.idg.controllers.IDGApp
@import ix.idg.controllers.Commons
@import java.net.URLEncoder

@(title: String)

<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
    <head>
        <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
        <meta http-equiv="X-UA-Compatible" content="IE=edge">

        <title>@title</title>
        <link rel="stylesheet"
        href='@routes.WebJarAssets.at(WebJarAssets.locate("bootstrap.min.css"))'>
        <link rel="stylesheet" type="text/css"
        href='@routes.WebJarAssets.at(WebJarAssets.locate("font-awesome.min.css"))'>
        <link rel="stylesheet" type="text/css"
        href='@routes.Assets.at("ncats/css/idg-landing-hex.css")'>
        <link rel="icon" type="image/png" href='@routes.Assets.at("images/favicon.png")'>

        <script type="text/javascript" src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

            <!--[if lte IE 9]>
<style type="text/css">
#ie6msg {border:3px solid #c33; margin:8px 0; background:#fcc; color:#000;}
#ie6msg h4{margin:8px; padding:0;}
#ie6msg p{margin:8px; padding:0;}
#ie6msg p a.getie7{font-weight:bold; color:#006;}
#ie6msg p a.ie6expl{font-weight:bold; color:#006;}
</style>
<div id="ie6msg">
<h4>Did you know that your browser is out of date, and potentially poses a security risk?</h4>
<p>To get the best possible experience using this website it is recommended that you upgrade your browser to a newer version.
<p>You could also try some other popular browsers like <a class="ie6expl" href="http://mozilla.com">FireFox</a> or
 <a class="ie6expl" href="https://www.google.com/chrome/S">Chrome</a>.</p>
</div>
<![endif]-->
        <script src='@routes.WebJarAssets.at(WebJarAssets.locate("jquery.min.js"))'
        type='text/javascript'></script>
        <script src='@routes.WebJarAssets.at(WebJarAssets.locate("bootstrap.min.js"))'></script>
        <script src='@routes.WebJarAssets.at(WebJarAssets.locate("typeahead.bundle.min.js"))'></script>
        <script src='@routes.WebJarAssets.at(WebJarAssets.locate("handlebars.min.js"))'></script>
        <script src='@routes.WebJarAssets.at(WebJarAssets.locate("bloodhound.min.js"))'></script>
        <script>
        ( function ( i, s, o, g, r, a, m ) { i[ 'GoogleAnalyticsObject' ] = r ; i[ r ] = i[ r ] || function ( ) {
        ( i[ r ].q = i[ r ].q ||[ ] ).push ( arguments ) }, i[ r ].l = 1 * new Date ( ) ; a = s.createElement ( o ),
        m = s.getElementsByTagName ( o )[ 0 ] ; a.async = 1 ; a.src = g ; m.parentNode.insertBefore ( a, m )
        } ) ( window, document, 'script', '//www.google-analytics.com/analytics.js', 'ga' ) ;

        ga ( 'create', 'UA-53774738-1', 'auto' ) ;
        ga ( 'send', 'pageview' ) ;
        </script>
    </head>
    <body>
        <div class="container vertical-center" id="main-box">
            <div class="row">

                <center>
                    <a href='@ix.idg.controllers.routes.IDGApp.index()'>
                        <img src='@routes.Assets.at("ncats/images/Pharos_Logo_Rd8.svg")' class = "img-responsive" id="pharos" />
                    </a>
                </center>
                <h2>Fractional counting of protein mentions</h2>
                <h3>Named entity recognition (NER)</h3>
                <p>The recognition of protein and gene names in text builds upon the text-mining pipeline of the STRING database
                    (Franceschini et al., 2013). This relies heavily on a highly efficient and flexible NER engine, which is
                    implemented in C++ and is described in full detail in Pafilis et al. (2013). Most recently we have
                    updated the dictionaries and underlying human genome annotation as part of releasing STRING version 10
                    (Szklarczyk et al., 2015). The dictionary merges synonym information from multiple sources, including
                    the Ensembl (Cunningham et al., 2015) and UniProt (UniProt Consortium, 2015) databases.</p>

                <h3>Text corpora</h3>
                <p>To construct a literature corpus, we first downloaded all articles from the PubMed Central (PMC)
                    <a href="http://www.ncbi.nlm.nih.gov/pmc/tools/openftlist/">Open Access Subset</a> and converted
                    them to the format required by the NER software. We next queried PubMed for additional articles for
                    which full-text is freely available online and automatically downloaded these in PDF format from
                    the publisher’s website when possible. We used <a href="http://www.abiword.org/">AbiWord</a> to
                    convert these to HTML format and subsequently converted this to the format required by the
                    NER software. Because the text is extracted from PDF files, the text is subject to formatting
                    artefacts. The resulting combined corpus consists of approximately 2 million full-text articles.
                    Despite this effort, we presently do not have access to mine the full-text articles for the majority
                    of Medline entries. We thus extend the literature corpus with these abstracts, which we extracted
                    from a local copy of <a href="http://www.nlm.nih.gov/databases/journal.html">Medline</a> and
                    converted to the same format as the full-text articles.</p>
                <p>We complement the literature corpus with a second corpus of USPTO patent texts, which we downloaded
                    from <a href="http://www.google.com/googlebooks/uspto-patents-grants-text.html">Google Books</a>.
                    As the file format of these have changed over the years, we developed several
                    parsers to convert them all into the same unified format as the literature corpus. Most notably,
                    all patents prior to 2002 were scanned and converted to plain text through optical character
                    recognition (OCR); these are thus also subject to OCR errors.</p>

                <h3>Fractional counting</h3>
                <p>A document can mention multiple proteins without pertaining equally much to all of them. To address
                    this, we use a fractional counting scheme in which each paper that mentions at least one protein
                    contributes a total count of 1, which is distributed across the mentioned proteins relative to
                    how many times each of them was mentioned. The total count for protein i is thus:</p>
                $$N_i = \sum_{j \in D} \frac{n_{ij}}{n_j}$$
                <p>where \(D\) is the document set, \(n_{ij}\) is the number of times protein \(i\) is mentioned in
                    document \(j\), \(n_j\) is total number of mentions of any protein in document \(j\).</p>


                <h2>References</h2>
                <p>Cunningham F, Amode MR, Barrell D, Beal K, Billis K, Brent S, Carvalho-Silva D, Clapham P, Coates G,
                    Fitzgerald S, Gil L, Girón CG, Gordon L, Hourlier T, Hunt SE, Janacek SH, Johnson N, Juettemann T,
                    Kähäri AK, Keenan S, Martin FJ, Maurel T, McLaren W, Murphy DN, Nag R, Overduin B, Parker A,
                    Patricio M, Perry E, Pignatelli M, Riat HS, Sheppard D, Taylor K, Thormann A, Vullo A, Wilder SP,
                    Zadissa A, Aken BL, Birney E, Harrow J, Kinsella R, Muffato M, Ruffier M, Searle SM, Spudich G,
                    Trevanion SJ, Yates A, Zerbino DR, Flicek P (2015). <a href="http://dx.doi.org/10.1093/nar/gku1010">
                    Ensembl 2015</a>. Nucleic Acids Research,
                    43:D662–D669</p>
                <p>Franceschini A, Szklarczyk D, Frankild S, Kuhn M, Simonovic M, Roth A, Lin J, Minguez P, Bork P,
                    von Mering C, Jensen LJ (2013). <a href="http://dx.doi.org/10.1093/nar/gks1094">
                    STRING v9.1: protein-protein interaction networks, with increased
                    coverage and integration</a>. Nucleic Acids Research, 41:D808–D815</p>

                <p>Pafilis E, Frankild SP, Fanini L, Faulwetter S, Pavloudi C, Vasileiadou A, Arvanitidis C, Jensen LJ
                    (2013). <a href="http://dx.doi.org/10.1371/journal.pone.0065390">The SPECIES and ORGANISMS resources
                    for fast and accurate identification of taxonomic names
                    in text</a>. PLOS ONE, 8:e65390</p>


                <p>Szklarczyk D, Franceschini A, Wyder S, Forslund K, Heller D, Huerte-Cepas J, Simonovic M, Roth A,
                    Santos A, Tsafou KP, Kuhn M, Bork P, Jensen LJ, von Mering C (2015).
                    <a href="http://dx.doi.org/10.1093/nar/gku1003">STRING v10: protein–protein
                    interaction networks, integrated over the tree of life</a>. Nucleic Acids Research, 43:D447–D452.
                </p>

                <p>UniProt Consortium (2015). <a href="http://dx.doi.org/10.1093/nar/gku989">UniProt: a hub for protein
                    information</a>. Nucleic Acids Research,
                    43:D204–D212</p>

            </div>
    </body>
</html>
